### PR TITLE
migrate scatter_ to ATen CPU (+multithreading, nondeterministic)

### LINKED ISSUE #24757: Migrate `scatter` and `scatter_` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by nikitaved**
Fixes #24757, partially #33094. Uses fix introduces in #33108 to avoid regressions for some compilers.

### CODE REVIEW

**Comment by nikitaved** [comment]
@ezyang, could you please help with the CI issues? All the tests from above pass on my machine (`torch_nn, jit_simple, autograd`). Shall I return the `scatter_.value` dispatch back?

**Comment by VitalyFedyunin** [comment]
Please add benchmarks. Also if possible please measure the speed of the deterministic variant with `/*serial_exec=*/true`

**Comment by ezyang** [comment]
It looks like the Scalar and Tensor overloads are a package deal. You have to port them both at once, or keep the old code around so the Scalar overload keeps working.

**Comment by nikitaved** [comment]
@ezyang, yes, this is what I thought. But then how come all the test that accept `Scalar` instead of `Tensor` pass on my machine and some other CIs even though the responsible TH code is removed?

**Comment by ngimel** [comment]
> Please add benchmarks. Also if possible please measure the speed of the deterministic variant with `/*serial_exec=*/true`

I think `scatter` should follow numpy semantics for advanced assignment (because `scatter` is a subset of what can be expressed with advanced assignment in numpy), and there is not guarantee of iteration order in this case:
```
For advanced assignments, there is in general no guarantee for the iteration order. This means that if an element is set more than once, it is not possible to predict the final result.
```
https://numpy.org/doc/1.18/reference/arrays.indexing.html

File: aten/src/ATen/native/cpu/ScatterGatherKernel.cpp [hunk 1]
```diff
@@ -72,9 +72,34 @@ void scatter_shape_check(
...
+    " to be smaller size than src ", src.sizes(),
+    " and to be smaller than self ", self.sizes(),
+    " apart from dimension ", dim
+  );
+}
+
+// Used for `scatter` when source is Scalar
+// Tests:
+//  1. index.size(d) <= self.size(d) for all d != dim
+void scatter_shape_check(
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/ScatterGatherKernel.cpp]: `scatter_shape_check` overloads are confusing and repetitive, can they be handled in the same function with an optional `src` argument?

**Comment by nikitaved** [comment]
@ngimel, there is something strange happening with this PR which is a bit beyond my understanding. More specifically, it does not seem to parallelize for some reason...

Let's take a step back and test whether parallelization actually works using `gather` operation. 
I am using AMD Ryzen Threadripper 2970WX and the following sparse indexing test:
```python
import torch
from IPython import get_ipython

ipython = get_ipython()

#torch.set_num_threads(1)

nrows = 50000
ncols = 100000
dims = [nrows, ncols]

src = torch.randn(dims)
idx1 = torch.randint(dims[0], (1, dims[1])).long()
idx2 = torch.randint(dims[1], (dims[0], 1)).long()

ipython.magic("timeit src.gather(0, idx1)")
ipython.magic("timeit src.gather(1, idx2)")
```

I get the following timings with the test:
gather ATen:
```
torch.set_num_threads(1)
4.18 ms ± 32.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)                                                                           
1.52 ms ± 13.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

#torch.set_num_threads(1)
534 µs ± 83.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
268 µs ± 1.56 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
gather TH:
```
torch.set_num_threads(1)
21.4 ms ± 5.11 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
9.55 ms ± 142 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

#torch.set_num_threads(1)
21.3 ms ± 14.1 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
9.56 ms ± 68.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

I slightly modify the test script from above to have the same complexity and try it with `scatter`:
```python
import torch
from IPython import get_ipython

ipython = get_ipython()

torch.set_num_threads(1)

nrows = 50000
ncols = 100000
dims = [nrows, ncols]

res = torch.randn(dims)
idx1 = torch.randint(dims[0], (1, dims[1])).long()
src1 = torch.randn(1, dims[1])
idx2 = torch.randint(dims[1], (dims[0], 1)).long()
src2 = torch.randn(dims[0], 1)

ipython.magic("timeit res.scatter(0, idx1, src1)")
ipython.magic("timeit res.scatter(1, idx2, src2)")

```

scatter ATen:
```
torch.set_num_threads(1)
4.92 s ± 461 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                
4.52 s ± 103 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) 

#torch.set_num_threads(1)
2.24 s ± 55.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
2.38 s ± 211 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

scatter TH:
```
torch.set_num_threads(1)
4.73 s ± 306 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                          
4.56 s ± 124 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) 

#torch.set_num_threads(1)
2.25 s ± 59.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
2.24 s ± 40.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```
The grain size used by TensorIterator is roughly 35k, so I have no idea why I observe this, unless I am missing something...

**Comment by nikitaved** [comment]
Ah, ok, looks like I should be using `scatter_` instead of `scatter`.

**Comment by nikitaved** [comment]
@ngimel, would you like me to perform some benchmarks like similar to what we did with `gather`? Other that that it seems like this PR is ready for review.

**Comment by ngimel** [comment]
Yeah, benchmarking would be useful, probably with the same sizes. Thanks!

**Review by ngimel — APPROVED** [review]
Approving, pending benchmarking results

**Comment by nikitaved** [comment]
@ngimel , these are the results. The testing scenarios are very similar to the ones we did for `gather`.
### Many-to-one case
```python
nfeat = 50
nrep = 2
a=torch.arange(nfeat).repeat_interleave(nrep)
batch=3 #batch can vary 1-200

res = torch.randn(1000, 1000)

for batch in [5, 10, 20, 50, 100, 200]:
    print("Batch: ", batch)
    c=torch.randint(3, (batch, nfeat * nrep)).float()
    ipython.magic("timeit res.scatter_(1,a.unsqueeze(0).expand(batch,a.size(0)),c)")

    enum_values = [
        list(range(1, 201)),
        list(range(1000, 1020)),
        list(range(2000, 2010)),
        list(range(3000, 3206)),
    ]
    indices = torch.tensor([i for i, values in enumerate(enum_values) for _j in range(len(values))])
    c = torch.randint(3, (batch, 4)).float()
    idx = indices.unsqueeze(0).expand(c.size(0), indices.size(0))
    src = c.repeat(1, idx.shape[-1] // c.shape[-1])
    ipython.magic("timeit res.scatter_(1,idx,src)")
    print()
```
#### ATen
```
Batch:  5
6.61 µs ± 60.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
3.99 µs ± 3.32 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  10
7.04 µs ± 27.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
5.32 µs ± 6.93 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  20
7.67 µs ± 11.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
7.99 µs ± 23.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  50
9.9 µs ± 195 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
15.9 µs ± 24.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  100
13.1 µs ± 176 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
29.1 µs ± 12.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

Batch:  200
20 µs ± 247 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
56.6 µs ± 595 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

```
#### TH
```
Batch:  5
5.51 µs ± 71.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
3.12 µs ± 6.58 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  10
6.33 µs ± 212 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
4.64 µs ± 4.65 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  20
7.55 µs ± 13 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
7.9 µs ± 53.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  50
12 µs ± 372 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
17.6 µs ± 447 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  100
18.5 µs ± 166 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
33 µs ± 71.5 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

Batch:  200
32.1 µs ± 78.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
64.7 µs ± 125 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

```
### Dense indexing
```python
Ms=256
Ns=512
dim = 0
top_power = 2

for pM in range(top_power):
    M = Ms * (2 ** pM)
    for pN in range(top_power):
        N = Ns * (2 ** pN)
        input_one = torch.rand(M, N)
        index = torch.tensor(numpy.random.randint(0, M, (M, N)))
        res = torch.randn(M, N)
        print(f"Problem size (MxN): {M}x{N}")
        ipython.magic("timeit res.scatter_(dim, index, input_one)")
```
#### ATen
```
Problem size (MxN): 256x512
547 µs ± 5.93 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 256x1024
1.33 ms ± 273 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 512x512
842 µs ± 33 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 512x1024
3 ms ± 20.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```
#### TH
```
Problem size (MxN): 256x512
570 µs ± 33.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 256x1024
1.13 ms ± 4.71 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 512x512
960 µs ± 8.74 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 512x1024
3.04 ms ± 11.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```

### Sparse indexing
```python
nrows = 3000
ncols = 10000
dims = [nrows, ncols]

res = torch.randn(dims)
idx1 = torch.randint(dims[0], (1, dims[1])).long()
src1 = torch.randn(1, dims[1])
idx2 = torch.randint(dims[1], (dims[0], 1)).long()
src2 = torch.randn(dims[0], 1)

ipython.magic("timeit res.scatter_(0, idx1, src1)")
ipython.magic("timeit res.scatter_(1, idx2, src2)")
```
#### ATen
```
50.3 µs ± 583 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
15.7 µs ± 24.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
#### TH
```
752 µs ± 9.54 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
217 µs ± 10.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

```

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**